### PR TITLE
[core] Clean up GCS

### PR DESCRIPTION
**Pull request by vitsai**
KV store and gcs table store use different underlying StoreClients, which then use the same underlying RedisClient. Clean it up so they use the same StoreClient, and also refactor a bunch of shared_ptrs to be unique_ptrs owned by GCS (with raw passed to non-owners). Also delete a couple now-useless GcsTableStorage classes.

## Why are these changes needed?

tech debt

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/store_client/redis_store_client.h [hunk 1]
```diff
@@ -110,7 +110,7 @@ class RedisStoreClient : public StoreClient {
     /// The pending shard scan count.
     std::atomic<size_t> pending_request_count_{0};
 
-    std::shared_ptr<RedisClient> redis_client_;
+    RedisClient *redis_client_;
```
Comment (iycheng) [inline src/ray/gcs/store_client/redis_store_client.h:113]: Should we make it a reference?

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 2]
```diff
@@ -135,7 +135,7 @@ void RedisStoreClient::MGetValues(const std::string &table_name,
   }
 }
 
-RedisStoreClient::RedisStoreClient(std::shared_ptr<RedisClient> redis_client)
+RedisStoreClient::RedisStoreClient(std::unique_ptr<RedisClient> &&redis_client)
```
Comment (iycheng) [inline src/ray/gcs/store_client/redis_store_client.cc:138]: I think `&&` is not necessary?

File: src/ray/gcs/gcs_server/store_client_kv.h [hunk 3]
```diff
@@ -58,7 +58,7 @@ class StoreClientInternalKV : public InternalKVInterface {
             std::function<void(std::vector<std::string>)> callback) override;
 
  private:
-  std::unique_ptr<StoreClient> delegate_;
+  StoreClient *delegate_;
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/store_client_kv.h:61]: Prefer reference.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 4]
```diff
@@ -72,17 +72,32 @@ GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
...
+        std::make_unique<InMemoryStoreClient>(main_service_));
     break;
   case StorageType::REDIS_PERSIST:
-    gcs_table_storage_ = std::make_shared<gcs::RedisGcsTableStorage>(GetOrConnectRedis());
+    auto redis_client = std::make_unique<RedisClient>(GetRedisClientOptions());
+    auto status = redis_client->Connect(main_service_);
+    RAY_CHECK(status.ok()) << "Failed to init redis gcs client as " << status;
+
+    // Init redis failure detector.
+    gcs_redis_failure_detector_ = std::make_unique<GcsRedisFailureDetector>(
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:90]: Should we move this into the RedisClient? (Ok not doing it here).

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 5]
```diff
@@ -555,20 +570,8 @@ void GcsServer::InitUsageStatsClient() {
 
 void GcsServer::InitKVManager() {
   // TODO (yic): Use a factory with configs
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:572]: Remove?

**Review by fishbone — COMMENTED** [review]
I think the direction looks fine. My only concern is about performance. Previously it has two tables (in-memory) and now it only has one.

Could you run actor stress tests and many actor nodes stress tests to see whether there is any regression?

File: src/ray/gcs/store_client/redis_store_client.cc (hunk 2 above)
Comment (vitsai) [reply src/ray/gcs/store_client/redis_store_client.cc]: Ah, I see what you mean, unique_ptr can't be copied anyway so no need

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 4 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.cc]: Did it in the latest commit

File: src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc [hunk 6]
```diff
@@ -143,7 +143,7 @@ class GcsActorManagerTest : public ::testing::Test {
 
     gcs_publisher_ = std::make_shared<gcs::GcsPublisher>(std::move(publisher));
     store_client_ = std::make_shared<gcs::InMemoryStoreClient>(io_service_);
-    gcs_table_storage_ = std::make_shared<gcs::GcsTableStorage>(store_client_.get());
+    gcs_table_storage_ = std::make_shared<gcs::GcsTableStorage>(*store_client_.get());
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/test/gcs_actor_manager_test.cc:146]: nit, you can just deference the store_client_ right?

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 7]
```diff
@@ -135,12 +136,18 @@ void RedisStoreClient::MGetValues(const std::string &table_name,
   }
 }
 
-RedisStoreClient::RedisStoreClient(std::unique_ptr<RedisClient> &&redis_client)
+RedisStoreClient::RedisStoreClient(std::unique_ptr<RedisClient> redis_client,
+                                   instrumented_io_context &io_context)
     : external_storage_namespace_(::RayConfig::instance().external_storage_namespace()),
-      redis_client_(std::move(redis_client)) {
+      redis_client_(std::move(redis_client)),
+      failure_detector_(io_context, *redis_client_.get(), []() {
```
Comment (iycheng) [inline src/ray/gcs/store_client/redis_store_client.cc:143]: just deference the redis_client_?

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 8]
```diff
@@ -180,7 +187,7 @@ Status RedisStoreClient::AsyncGetAll(
   std::string match_pattern =
       GenKeyRedisMatchPattern(external_storage_namespace_, table_name);
   auto scanner = std::make_shared<RedisScanner>(
-      redis_client_.get(), external_storage_namespace_, table_name);
+      *redis_client_.get(), external_storage_namespace_, table_name);
```
Comment (iycheng) [inline src/ray/gcs/store_client/redis_store_client.cc:190]: ditto

**Review by fishbone — COMMENTED** [review]
LG. 👍

**Comment by vitsai** [comment]
Oh nice release tests passed: https://buildkite.com/ray-project/release-tests-pr/builds/52702

**Comment by iycheng** [comment]
@vitsai  could you compare the metrics there? These are perf tests, so metrics are important.

**Comment by iycheng** [comment]
cpp test failed to build.

**Comment by vitsai** [comment]
Going to rebase and rerun release tests/microbenchmarks

**Comment by vitsai** [comment]
[Microbenchmarks on master](https://buildkite.com/ray-project/release-tests-branch/builds/1592#0187a4b4-b2bb-44c0-889f-22f00002a778): 

single_client_get_calls_Plasma_Store = [6220.115319914286, 224.5052300173086]
--
  | single_client_put_calls_Plasma_Store = [6427.78729348863, 84.21488331092435]
  | multi_client_put_calls_Plasma_Store = [13685.622603708454, 107.89607650269706]
  | single_client_put_gigabytes = [20.114238761619227, 6.003066360606937]
  | single_client_tasks_and_get_batch = [10.621044232599615, 0.7018065234293067]
  | multi_client_put_gigabytes = [31.944310601566972, 0.8025680665642678]
  | single_client_get_object_containing_10k_refs = [13.4303118492593, 0.17122469816034125]
  | single_client_wait_1k_refs = [5.669743175103769, 0.16404015686449241]
  | single_client_tasks_sync = [1402.5799311893395, 44.22066668152743]
  | single_client_tasks_async = [11589.713176381527, 69.968115451784]
  | multi_client_tasks_async = [34377.35783189367, 2098.212516049616]
  | 1_1_actor_calls_sync = [2627.6907826949946, 17.95559542036301]
  | 1_1_actor_calls_async = [8774.565215109124, 99.25145099727403]
  | 1_1_actor_calls_concurrent = [5556.408295085073, 310.93719458837893]
  | 1_n_actor_calls_async = [11443.027888783963, 87.49946447311738]
  | n_n_actor_calls_async = [34184.700321977085, 833.4165939251417]
  | n_n_actor_calls_with_arg_async = [3086.29057625603, 35.752775132663736]
  | 1_1_async_actor_calls_sync = [1749.332379578524, 36.34387252208381]
  | 1_1_async_actor_calls_async = [3053.0727243465535, 44.040906553833615]
  | 1_1_async_actor_calls_with_args_async = [2438.8474219503923, 88.20268908351537]
  | 1_n_async_actor_calls_async = [10589.09103935827, 106.34051316384893]
  | n_n_async_actor_calls_async = [27000.281282494278, 1192.3434495510094]
  | placement_group_create/removal = [1111.4458914419295, 36.12535104191624]
  | client__get_calls = [1150.072552279968, 36.79667266684934]
  | client__put_calls = [904.8795868787086, 13.073088078601502]
  | client__put_gigabytes = [0.045687216636994404, 0.00042081305098886794]
  | client__tasks_and_put_batch = [12964.98005883289, 319.3993930878376]
  | client__1_1_actor_calls_sync = [569.89256256793, 24.8044284674939]
  | client__1_1_actor_calls_async = [1084.4466197839831, 33.13091505679245]
  | client__1_1_actor_calls_concurrent = [1106.1285553207586, 25.829782511660305]
  | client__tasks_and_get_batch = [0.9467558674435517, 0.053000735219486415]

**Pull request closed without merging**